### DISCUSSION TITLE
Meaning of geo.x in osc.lua file

### DISCUSSION THREAD

**Discussion by advice2020** [mpv-player/mpv#14915] [General]
I have been trying to make some modifications to the “bottombar” layout in the osc.lua file.
I am trying to figure out how to center the default timecode + seekbar + timecode but I am having trouble doing so.
I was wondering does anyone know what / where the "geo." items originate from?
Such as geo.x  &  geo.y
I am assuming that geo = geometry and x & y are coordinates but I do not know where these originate from / what are they referencing?

Here are some code examples that use them......
```
-- Left timecode
    geo = { x = geo.x + geo.w + padX + tcW, y = geo.y, an = 6,
            w = tcW, h = geo.h }
            
-- Right timecode
    geo = { x = geo.x - padX - tcW - 10, y = geo.y, an = geo.an,
            w = tcW, h = geo.h }
            
-- Seekbar
    geo = { x = sb_l, y = geo.y, an = geo.an,
            w = math.max(0, sb_r - sb_l), h = geo.h }       
```     
Does anyone have any information they can share about these "geo" based items?

Thank You

**Comment by advice2020** [comment]
Still interested if anyone knows about this or where I can find information about something like this.
Thank You

**Comment by Samillion** [comment]
If the OSC you're working with is following the usual standard, before that code you should find a `local geo` list, something like:

```lua
    local geo = {
        x = 0,
        y = 0,
        an = 1,
        w = 0,
        h = 0,
    }
```

> I am assuming that geo = geometry and x & y are coordinates but I do not know where these originate from / what are they referencing?

That is correct. Think of them as measurements to assign where the element would be from all sides.

If I may suggest something, to not re-invent the wheel, since you want it centered, I'd recommend checking the official `osc.lua`, specifically the box layout.

It is essentially what you want, centered. You can then adjust, add or remove elements as you like, and it is well commented for you to understand what each variable or function does. For example, from `osc.lua`:

```lua
-- Classic box layout
layouts["box"] = function ()

    local osc_geo = {
        w = 550,    -- width
        h = 138,    -- height
        r = 10,     -- corner-radius
        p = 15,     -- padding
    }
```

**Comment by advice2020** [comment]
@Samillion 
Thank you again for your responses (I apologize all the responses back to you :)
I believe I am aware of the things you have mentioned, all the items you are mentioning seem to be based on the other "layouts" MPV make available.  I had looked into the other "layouts" before starting this but they are all overlays on top which I did not want.  The tricky part is that the layout I am using is the "bottombar" which in the osc.lua file for some reason is not laid out the same as those other layouts and its coding is not as straight forward to me.  
It is not in the `-- Layouts` (comment) section with all the other layouts, it is in its own category below it called
`local function bar_layout(direction)`
Took me a long time to figure this out because it does not contain any kind of "commented" label stating that it is the "bottombar".
Here is the beginning of it as mentioned in my other post #14871
```
local function bar_layout(direction)
    local osc_geo = {
        x = -2,
        y = nil,
        an = (direction < 0) and 7 or 1,
        w = nil,
        h = 56,
    }
```
Confusing because of the "nil" values and the "an" values? not sure why done this way/what these mean?

So all the items that say "geo" such as geo.x and geo.y are referring to this this "local osc_geo = "
I was looking for/expecting something like what you said "local = geo" because I thought the "geo" label had to be exact to "osc_geo"

The question I have though is all the items that use these "geo" references all just say geo.x geo.y
So if all the elements use the x = geo.x that would mean they would all end up on top of each other.
My discovery was that I think all items are chained off of each other probably based on top to bottom order in the code.  There is also padding values which are probably the spaces between them.  Hopefully this makes sense how I am explaining it.
For example mentioned in my other post you saw that timecodes and seekbar by default are not centered.
I could not figure out how to move the timecodes so I kept my left timecode where it shows up by default and then I figured out how to change the seekbar width, when changing its width then the right timecode automatically moved over to the right because it was based off of the seekbar which was the item before it.

Anyways, again I apologize for all the replies, hard to explain all of this.
If I can just change the height of the black background and get all the items that get shifted down to be centered again I think my OSC would be okay to use.
If you or anyone else has any ideas, please let me know.
Thank you again for all your replies though.
